src/uri.rs

```diff
@@ -102,86 +112,90 @@ impl FromStr for UUri {
+    // [impl->dsn~uri-scheme~1]
+    // [impl->dsn~uri-host-only~1]
+    // [impl->dsn~uri-authority-mapping~1]
+    // [impl->dsn~uri-path-mapping~1]
+    // [impl->req~uri-serialization~1]
     fn from_str(uri: &str) -> Result<Self, Self::Err> {
         if uri.is_empty() {
             return Err(UUriError::serialization_error("URI is empty"));
         }
+        let parsed_uri = URIReference::try_from(uri)
+            .map_err(|e| UUriError::serialization_error(e.to_string()))?;
 
-        let uri_to_parse = uri.find(':').map_or_else(
-            || Ok(uri.replace('\\', "/")),
-            |_index| {
-                // strip leading scheme definition (`up`) up to and including `:`
-                uri.strip_prefix("up:")
-                    .ok_or(UUriError::serialization_error(
-                        "uProtocol URI must use 'up' scheme",
-                    ))
-                    .map(|s| s.to_string())
-                // if uri.starts_with("up:") {
-                //     // strip leading scheme definition (`up`) up to and including `:`
-                //     Ok(uri[3..].to_string())
-                // } else {
-                //     Err(UUriError::serialization_error(
-                //         "uProtocol URI must use 'up' scheme",
-                //     ))
-                // }
-            },
-        )?;
-        let is_local: bool = !uri_to_parse.starts_with("//");
-        let uri_parts = uri_to_parse.split('/').collect::<Vec<&str>>();
-
-        let mut authority_name: String = String::default();
-        #[allow(unused_assignments)]
-        let mut entity: String = String::default();
-        #[allow(unused_assignments)]
-        let mut version: String = String::default();
-        #[allow(unused_assignments)]
-        let mut resource: String = String::default();
-
-        if is_local {
-            if uri_parts.len() != 4 {
+        if let Some(scheme) = parsed_uri.scheme() {
+            if scheme.ne("up") {
                 return Err(UUriError::serialization_error(
-                    "Local URI must contain entity ID, entity version and resource ID",
+                    "uProtocol URI must use 'up' scheme",
                 ));
             }
-            entity = uri_parts[1].to_string();
-            version = uri_parts[2].to_string();
-            resource = uri_parts[3].to_string();
-        } else if uri_parts.len() != 6 {
+        }
+        if parsed_uri.has_query() {
+            return Err(UUriError::serialization_error(
+                "uProtocol URI must not contain query",
+            ));
+        }
+        if parsed_uri.has_fragment() {
             return Err(UUriError::serialization_error(
-                "Remote URI must contain authority, entity ID, entity version and resource ID",
+                "uProtocol URI must not contain fragment",
             ));
-        } else if uri_parts[2].trim().is_empty() {
+        }
+        let authority_name = parsed_uri
```

Comment:
It suddenly comes to my mind: Is it possible to have the asterisk in the authority name?
I mean `up//*/...` means matching all authority name, but how about `up/auth*/...`?

Reply:
It is (currently) possible to use an authority that includes an asterisk, but that would have no special meaning, i.e. it would not be recognized nor used as a wildcard when matching.